docarray/array/array/pushpull/__init__.py

```diff
@@ -0,0 +1,244 @@
+        url: str,
+        public: bool,
+        show_progress: bool,
+        branding: Optional[Dict],
+    ) -> Dict:
+        ...
+
+    @staticmethod
+    def push_stream(
+        docs: Iterator['BaseDocument'],
+        url: str,
+        public: bool = True,
+        show_progress: bool = False,
+        branding: Optional[Dict] = None,
+    ) -> Dict:
+        ...
+
+    @staticmethod
+    def pull(
+        cls: Type['DocumentArray'],
+        name: str,
+        show_progress: bool,
+        local_cache: bool,
+    ) -> 'DocumentArray':
+        ...
+
+    @staticmethod
+    def pull_stream(
+        cls: Type['DocumentArray'],
+        name: str,
+        show_progress: bool,
+        local_cache: bool,
+    ) -> Iterator['BaseDocument']:
+        ...
+
+
+class PushPullMixin(Iterable['BaseDocument']):
+    """Mixin class for push/pull functionality."""
+
+    __backends__: Dict[str, PushPullLike] = {}
+    document_type: Type['BaseDocument']
+
+    @abstractmethod
+    def __len__(self) -> int:
+        ...
+
+    @classmethod
+    def get_backend(cls, protocol: str) -> PushPullLike:
+        """
+        Get the backend for the given protocol.
+
+        :param protocol: the protocol to use, e.g. 'jinaai', 'file', 's3'
+        :return: the backend class
+        """
+        if protocol in cls.__backends__:
+            return cls.__backends__[protocol]
+
+        if protocol == 'jinaai':
+            from docarray.array.array.pushpull.jinaai import PushPullJAC
+
+            cls.__backends__[protocol] = PushPullJAC
+            logging.info('Loaded Jina AI Cloud backend')
+        elif protocol == 'file':
+            from docarray.array.array.pushpull.file import PushPullFile
+
+            cls.__backends__[protocol] = PushPullFile
+            logging.info('Loaded Local Filesystem backend')
+        elif protocol == 's3':
+            from docarray.array.array.pushpull.s3 import PushPullS3
+
+            cls.__backends__[protocol] = PushPullS3
+            logging.info('Loaded S3 backend')
+        else:
+            raise NotImplementedError(f'protocol {protocol} not supported')
+
+        return cls.__backends__[protocol]
+
+    @staticmethod
+    def list(
+        url: str = f'file://{__cache_path__}', show_table: bool = False
+    ) -> List[str]:
+        """
+        List all the DocumentArrays in the namespace.
+        url should be of the form ``protocol://namespace``
+
+        If no url is provided, the DocumentArrays in the local cache will be listed.
+
+        :param url: should be of the form ``protocol://namespace``. e.g. ``s3://bucket/path/to/namespace``, ``file:///path/to/folder``
+        :param show_table: whether to show the table of artifacts
+        :return: a list of artifact names
+        """
+        logging.info(f'Listing artifacts from {url}')
+        protocol, namespace = url.split('://', 2)
+        return PushPullMixin.get_backend(protocol).list(namespace, show_table)
```

Comment:
I would make it a class method here rather that staticmethod that called itself. This would make it easier if someone want to extend the behavior of the mixin without copy pasting code